cmd/nexusUpload.go

```diff
@@ -0,0 +1,256 @@
+package cmd
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"fmt"
+
+	"github.com/SAP/jenkins-library/pkg/command"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/maven"
+	"github.com/SAP/jenkins-library/pkg/nexus"
+	"github.com/SAP/jenkins-library/pkg/piperenv"
+	"github.com/SAP/jenkins-library/pkg/piperutils"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"github.com/ghodss/yaml"
+)
+
+func nexusUpload(options nexusUploadOptions, telemetryData *telemetry.CustomData) {
+	// error situations should stop execution through log.Entry().Fatal() call which leads to an os.Exit(1) in the end
+	runNexusUpload(&options, telemetryData)
+}
+
+func runNexusUpload(options *nexusUploadOptions, telemetryData *telemetry.CustomData) {
+	projectStructure := piperutils.ProjectStructure{}
+
+	nexusClient := nexus.Upload{Username: options.User, Password: options.Password}
+
+	if projectStructure.UsesMta() {
+		log.Entry().Info("MTA project structure detected")
+		uploadMTA(&nexusClient, options)
+	} else if projectStructure.UsesMaven() {
+		log.Entry().Info("Maven project structure detected")
+		uploadMaven(&nexusClient, options)
+	} else {
+		log.Entry().Fatal("Unsupported project structure")
+	}
+}
+
+func uploadMTA(nexusClient *nexus.Upload, options *nexusUploadOptions) {
+	if options.GroupID == "" {
+		log.Entry().Fatal("The 'groupID' parameter needs to be provided for MTA projects")
+	}
+	err := nexusClient.SetBaseURL(options.Url, options.Version, options.Repository, options.GroupID)
+	if err == nil {
+		err = setVersionFromMtaYaml(nexusClient)
+	}
+	if err == nil {
+		artifactID := options.ArtifactID
+		if artifactID == "" {
+			artifactID = piperenv.GetParameter(".pipeline/commonPipelineEnvironment/configuration", "artifactId")
+		}
+		err = nexusClient.AddArtifact(nexus.ArtifactDescription{File: "mta.yaml", Type: "yaml", Classifier: "", ID: options.ArtifactID})
+	}
+	if err == nil {
+		mtarFilePath := piperenv.GetParameter(".pipeline/commonPipelineEnvironment", "mtarFilePath")
+		fmt.Println(mtarFilePath)
+		err = nexusClient.AddArtifact(nexus.ArtifactDescription{File: mtarFilePath, Type: "mtar", Classifier: "", ID: options.ArtifactID})
+	}
+	if err == nil {
+		err = nexusClient.UploadArtifacts()
+	}
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
```

Comment:
let's bubble up the error instead of having the `Fatal()` on the lower level.
Testing will also likely be easier ...